### PR TITLE
perf: add Moq1208 benchmarks for issue #777 completion

### LINKED ISSUE #777: Tasks for Issue #767: Detect Returns() with Non-Task Delegate on Async Methods
## Parent Issue

Related to: #767 (Explainer and PRD)

## Overview

This task list tracks the implementation of a new Moq analyzer (Moq12XX) to detect when `.Returns(() => value)` is used with an incorrect delegate signature on async methods, causing runtime exceptions.

**Corrected Problem Statement:**
- Using `.Returns(() => value)` where delegate returns non-Task type on methods returning `Task<T>`/`ValueTask<T>`
- Causes runtime exception: `"Invalid callback. Setup on method with return type 'Task' cannot invoke callback with return type 'int'"`
- Root cause: Moq's catch-all `IReturns<TMock, TResult>.Returns(Delegate valueFunction)` signature

**Note:** The original explainer mentioned `SetupAsync()` which does NOT exist in Moq. The real problem is the delegate type mismatch in `.Returns()`.

## Relevant Files

Files that will need to be created or modified:

- `src/Analyzers/Moq12XXAnalyzer.cs` - Main analyzer implementation for detecting delegate type mismatch
- `src/CodeFixes/Moq12XXCodeFix.cs` - Code fix to transform Returns → ReturnsAsync
- `tests/Moq.Analyzers.Test/Moq12XXAnalyzerTests.cs` - Unit tests for analyzer detection logic
- `tests/Moq.Analyzers.Test/Moq12XXCodeFixTests.cs` - Unit tests for code fix transformations
- `tests/Moq.Analyzers.Benchmarks/Moq12XXBenchmarks.cs` - Performance benchmarks
- `docs/rules/Moq12XX.md` - Rule documentation with examples and rationale
- `src/Analyzers/AnalyzerReleases.Unshipped.md` - Release notes entry for new analyzer
- `README.md` - Update analyzer list and badges

## Notes

When editing files, follow the guidance at `.github/instructions/README.md` to determine appropriate instructions for specific files.

### Key Technical Constraints

- **Character-precise diagnostic spans** (non-negotiable requirement)
- **IOperation-based analysis** (follow Moq1203 pattern)
- **Data-driven tests** with `.WithMoqReferenceAssemblyGroups()`
- **Version compatibility**: Moq 4.8.2, 4.16.0+, 4.18.4+
- **No trial-and-error**: Must understand Roslyn APIs before implementing

### Success Criteria

- 95%+ true positive rate (correctly identifies delegate type mismatch)
- <5% false positive rate
- 100% code fix success rate
- <100ms analysis time per file
- Zero test failures

### Estimated Effort

- Total: 16-24 hours (per triage analysis)
- Analyzer: 8-12 hours
- Code fix: 4-6 hours
- Documentation: 4-6 hours

## High-Level Tasks

Below are the main phases of work required to implement this analyzer. Each will be broken down into detailed sub-tasks once you respond with "Go".

### Task 1: Project Setup and Planning
- Assign diagnostic ID (Moq12XX - next available in 1200-1299 range)
- Update issue #767 description to remove SetupAsync references
- Set up implementation branch
- Review existing similar analyzers (Moq1201, Moq1203, Moq1206)
- Identify reusable utilities from Common project

### Task 2: Analyzer Implementation
- Create Moq12XXAnalyzer.cs with IOperation-based detection
- Implement delegate return type inspection logic
- Handle Task<T>, Task, ValueTask<T>, ValueTask return types
- Calculate character-precise diagnostic spans
- Add version-aware logic for ValueTask (Moq 4.18.4+)

### Task 3: Code Fix Implementation
- Create Moq12XXCodeFix.cs using DocumentEditor pattern
- Transform `Returns` → `ReturnsAsync` while preserving delegate body
- Handle edge cases (chained calls, comments, formatting)
- Support FixAll provider for batch corrections
- Ensure 100% compilation success rate

### Task 4: Comprehensive Testing
- Create data-driven analyzer tests covering all scenarios
- Create data-driven code fix tests with before/after pairs
- Test all async return types (Task, Task<T>, ValueTask, ValueTask<T>)
- Test all Moq versions (4.8.2, 4.16.0+, 4.18.4+)
- Test edge cases (nested Tasks, generic methods, callbacks)
- Validate diagnostic span precision

### Task 5: Performance and Documentation
- Create performance benchmarks following Moq1206 pattern
- Validate <100ms analysis time requirement
- Write comprehensive rule documentation (docs/rules/Moq12XX.md)
- Update AnalyzerReleases.Unshipped.md
- Update README.md with new analyzer
- Validate code coverage meets requirements

---

**Status:** Ready for Phase 2 (detailed sub-task generation)

**Next Step:** Respond with "Go" to proceed with breaking down each high-level task into detailed, actionable sub-tasks.

### PR DESCRIPTION
**Pull request by rjmurillo**
## Summary

Completes Moq1208 (ReturnsDelegateShouldReturnTask) by adding benchmarks and removing dead code from the fixer.

Closes #777

## Changes

- Added `Moq1208ReturnsDelegateBenchmarks.cs` following the established benchmark pattern
- Removed unreachable `GenericNameSyntax` branch from `ReturnsDelegateShouldReturnTaskFixer.cs` (Moq has no `Returns<T>()` generic overload)
- Added `[BenchmarkCategory("Moq1208")]` and `[Benchmark(Baseline = true)]` attributes

## Performance

The Moq1208 benchmark is new (no baseline comparison available in CI). First baseline will be established on merge. For reference, comparable IOperation-based analyzers on the same CI runner (ARM64 Neoverse-N2):

| Analyzer | Diagnostic (1 file) | Baseline (1 file) | Ratio | Allocated |
|---|---|---|---|---|
| Moq1200 (setup overridable) | 2,218 us | 235 us | 9.5x | 215 KB |
| Moq1100 (callback signature) | 6,282 us | 236 us | 26.9x | 615 KB |
| Moq1000 (sealed class) | 5,564 us | 489 us | 11.8x | 151 KB |

Moq1208 performs similar work to Moq1200 (inspects invocation return types against the mocked method's return type). Expected overhead should be in the 2,000-6,000 us range per diagnostic file, with ~200 KB allocation. The baseline (empty analyzer) cost is ~235 us regardless of analyzer, dominated by Roslyn compilation setup.

**Key performance characteristics:**
- MoqKnownSymbols created per-compilation (not per-operation)
- Early exit via `IsMockReferenced()` when Moq is not referenced
- `EnableConcurrentExecution()` enabled
- Zero allocation in the hot path (symbol comparisons only)

## Dead code removal justification

The `GenericNameSyntax` branch in `ReplaceReturnsWithReturnsAsync` was unreachable:
1. Moq has no `Returns<T>()` generic overload
2. The analyzer's guard at line 45 (`ValueText == "Returns"`) ensures only `IdentifierName` nodes reach the method
3. Zero fixer tests exercised the branch

Three defensive null guards in `RegisterCodeFixesAsync` are retained per project convention (Roslyn CodeFix contract pattern).

## Test plan

- [x] Build passes with PedanticMode (0 warnings)
- [x] All 3,177 tests pass
- [x] Benchmark follows established pattern (IterationSetup, Baseline attribute)
- [x] No regressions detected by PerfDiff CI

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Release Notes

* **Tests**
  * Added comprehensive test coverage for generic `Returns<T>` patterns in mock setups with synchronous and asynchronous lambda scenarios.
  * Introduced performance benchmark suite to measure analyzer efficiency across varying configurations and dependency versions.

### CODE REVIEW

File: tests/Moq.Analyzers.Benchmarks/Moq1208ReturnsDelegateBenchmarks.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            .GetAnalysisResultAsync(CancellationToken.None)
+            .ConfigureAwait(false))
+            .AssertValidAnalysisResult()
+            .GetAllDiagnostics();
+
+        if (diagnostics.Length != 0)
+        {
+            throw new InvalidOperationException($"Expected no analyzer diagnostics but found '{diagnostics.Length}'");
+        }
+    }
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply tests/Moq.Analyzers.Benchmarks/Moq1208ReturnsDelegateBenchmarks.cs]: Addressed in 2399ec4:
1. `[IterationSetup]` is intentional and consistent with all other benchmarks in this repo (CompilationWithAnalyzers is stateful).
2. Added `[Benchmark(Baseline = true)]` and renamed to `Moq1208Baseline` to match convention.
3. A third "clean code" scenario is out of scope for this PR.
